File: source/includes/avs/bson-bindata-vectors/upload-create-index-existing-data.go
package main  
  
import (  
	"context"  
	"fmt"  
	"io/ioutil"  
	"log"  
    "time"
	"os"  
  
	"go.mongodb.org/mongo-driver/v2/bson"  
	"go.mongodb.org/mongo-driver/v2/mongo"  
	"go.mongodb.org/mongo-driver/v2/mongo/options"  
)  
  
var (  
	mongodbURI          = os.Getenv("MONGODB_URI")  
	dbName              = "sample_airbnb"  
	collectionName      = "listingsAndReviews"  
	indexName           = "<INDEX-NAME>"  
	numberOfDimensions  = 1024  
	embeddingFields     = []string{"embeddings_float32", "embeddings_int8", "embeddings_int1"}  
	embeddingSimilarity = []string{"dotProduct", "dotProduct", "euclidean"}  
)  
  
func main() {  
	if mongodbURI == "" {  
		log.Fatal("MONGODB_URI environment variable not set")  
	}  
  
	clientOptions := options.Client().ApplyURI(mongodbURI)  
	client, err := mongo.Connect(clientOptions)  
	if err != nil {  
		log.Fatalf("Error connecting to MongoDB: %v", err)  
	}  
	defer func() {  
		if err = client.Disconnect(context.TODO()); err != nil {  
			log.Fatal(err)  
		}  
	}()  
  
	if err := uploadEmbeddingsData(client); err != nil {  
		log.Fatalf("Error uploading embeddings data: %v", err)  
	}  
  
	setupVectorSearchIndex(client)  
}  
  
func uploadEmbeddingsData(client *mongo.Client) error {  
	collection := client.Database(dbName).Collection(collectionName)  
  
	// Load embeddings.json file  
	fileContent, err := ioutil.ReadFile("embeddings.json")  
	if err != nil {  
		return fmt.Errorf("error reading file: %w", err)  
	}  
  
	// Convert JSON file content to BSON compatible format using UnmarshalExtJSON  
	var documents []bson.M  
	if err := bson.UnmarshalExtJSON(fileContent, false, &documents); err != nil {  
		return fmt.Errorf("failed to unmarshal JSON data: %w", err)  
	}  
  
	// Update documents in MongoDB  
	for _, doc := range documents {  
		summary, exists := doc["text"].(string)  
		if !exists {  
			return fmt.Errorf("missing 'text' field in document")  
		}  
  
		// Using bson.Binary ensures binary data is correctly interpreted  
		if float32Bin, ok := doc["embeddings_float32"].(bson.Binary); ok {  
			doc["embeddings_float32"] = float32Bin  
		}  
		if int8Bin, ok := doc["embeddings_int8"].(bson.Binary); ok {  
			doc["embeddings_int8"] = int8Bin  
		}  
		if int1Bin, ok := doc["embeddings_int1"].(bson.Binary); ok {  
			doc["embeddings_int1"] = int1Bin  
		}  
  
		filter := bson.M{"summary": summary}  
		update := bson.M{  
			"$set": doc,  
		}  
  
		// Set the upsert option  
		opts := options.UpdateMany().SetUpsert(true)  
  
		_, err = collection.UpdateMany(context.TODO(), filter, update, opts)  
		if err != nil {  
			return fmt.Errorf("failed to update documents: %w", err)  
		}  
	}  
  
	return nil  
}  
  
// Sets up vector search index in MongoDB  
func setupVectorSearchIndex(client *mongo.Client) {  
	database := client.Database(dbName)  
	collection := database.Collection(collectionName)  
  
	ctx := context.TODO()  
  
	type vectorDefinitionField struct {  
		Type          string `bson:"type"`  
		Path          string `bson:"path"`  
		NumDimensions int    `bson:"numDimensions"`  
		Similarity    string `bson:"similarity"`  
	}  
  
	type vectorDefinition struct {  
		Fields []vectorDefinitionField `bson:"fields"`  
	}  
  
	fields := make([]vectorDefinitionField, len(embeddingFields))  
	for i, field := range embeddingFields {  
		fields[i] = vectorDefinitionField{  
			Type:          "vector",  
			Path:          field,  
			NumDimensions: numberOfDimensions,  
			Similarity:    embeddingSimilarity[i],  
		}  
	}  
  
	opts := options.SearchIndexes().SetName(indexName).SetType("vectorSearch")  
  
	indexModel := mongo.SearchIndexModel{  
		Definition: vectorDefinition{  
			Fields: fields,  
		},  
		Options: opts,  
	}  
  
	// Create the index  
	log.Println("Creating the index.")  
	searchIndexName, err := collection.SearchIndexes().CreateOne(ctx, indexModel)  
	if err != nil {  
		log.Fatalf("Failed to create the search index: %v", err)  
	}  
  
	// Polling to confirm successful index creation  
	log.Println("Polling to confirm successful index creation.")  
	log.Println("NOTE: This may take up to a minute.")  
	searchIndexes := collection.SearchIndexes()  
	var doc bson.Raw  
  
	for doc == nil {  
		cursor, err := searchIndexes.List(ctx, options.SearchIndexes().SetName(searchIndexName))  
		if err != nil {  
			log.Fatalf("failed to list search indexes: %v", err)  
		}  
  
		if !cursor.Next(ctx) {  
			break  
		}  
  
		name := cursor.Current.Lookup("name").StringValue()  
		queryable := cursor.Current.Lookup("queryable").Boolean()  
		if name == searchIndexName && queryable {  
			doc = cursor.Current  
		} else {  
			time.Sleep(5 * time.Second)  
		}  
	}  
  
	log.Println("Name of Index Created: " + searchIndexName)  
}  

File: content/atlas/source/includes/avs/bson-bindata-vectors/upload-create-index-existing-data.java
import com.mongodb.client.MongoClient;
import com.mongodb.client.MongoClients;
import com.mongodb.client.MongoCollection;
import com.mongodb.client.MongoDatabase;
import com.mongodb.client.model.SearchIndexModel;
import com.mongodb.client.model.SearchIndexType;

import org.bson.Document;
import org.bson.conversions.Bson;
import org.bson.BinaryVector; // Import the BinaryVector

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.TimeUnit;
import java.util.stream.StreamSupport;

public class UploadDataAndCreateIndex {

    private static final String MONGODB_URI = System.getenv("MONGODB_URI");
    private static final String DB_NAME = "sample_airbnb";
    private static final String COLLECTION_NAME = "listingsAndReviews";
    private static final String INDEX_NAME = "<INDEX-NAME>";

    public static void main(String[] args) {
        try (MongoClient mongoClient = MongoClients.create(MONGODB_URI)) {
            uploadEmbeddingsData(mongoClient);
            setupVectorSearchIndex(mongoClient);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public static void uploadEmbeddingsData(MongoClient mongoClient) throws IOException {
        MongoDatabase database = mongoClient.getDatabase(DB_NAME);
        MongoCollection<Document> collection = database.getCollection(COLLECTION_NAME);
        String filePath = "embeddings.json";
        String fileContent = Files.readString(Path.of(filePath));

        Document rootDoc = Document.parse(fileContent);
        List<Document> embeddingsDocs = rootDoc.getList("data", Document.class);

        for (Document doc : embeddingsDocs) {
            // Retrieve the string value from the document
            String summary = doc.getString("text");

            // Get the BinaryVector objects from the document
            BinaryVector embeddingsFloat32 = doc.get("embeddings_float32", BinaryVector.class);
            BinaryVector embeddingsInt8 = doc.get("embeddings_int8", BinaryVector.class);
            BinaryVector embeddingsInt1 = doc.get("embeddings_int1", BinaryVector.class);

            // Create filter and update documents
            Document filter = new Document("summary", summary);
            Document update = new Document("$set", new Document("summary", summary)
                    .append("embeddings_float32", embeddingsFloat32)
                    .append("embeddings_int8", embeddingsInt8)
                    .append("embeddings_int1", embeddingsInt1));

            // Perform update operation with upsert option
            collection.updateOne(filter, update, new com.mongodb.client.model.UpdateOptions().upsert(true));
            System.out.println("Processed document with summary: " + summary);
        }
    }

    public static void setupVectorSearchIndex(MongoClient client) throws InterruptedException {
        MongoDatabase database = client.getDatabase(DB_NAME);
        MongoCollection<Document> collection = database.getCollection(COLLECTION_NAME);
        // Define the index details
        Bson definition = new Document(
            "fields",
            List.of(
                new Document("type", "vector")
                    .append("path", "embeddings_float32")
                    .append("numDimensions", 1024)
                    .append("similarity", "dotProduct"),
                new Document("type", "vector")
                    .append("path", "embeddings_int8")
                    .append("numDimensions", 1024)
                    .append("similarity", "dotProduct"),
                new Document("type", "vector")
                    .append("path", "embeddings_int1")
                    .append("numDimensions", 1024)
                    .append("similarity", "euclidean")
            )
        );
        // Define the index model
        SearchIndexModel indexModel = new SearchIndexModel(
            INDEX_NAME,
            definition,
            SearchIndexType.vectorSearch()
        );
        // Create the index using the defined model
        List<String> result = collection.createSearchIndexes(Collections.singletonList(indexModel));
        System.out.println("Successfully created vector index named: " + result.get(0));
        System.out.println("It may take up to a minute for the index to leave the BUILDING status and become queryable.");
        // Wait for Atlas to build the index
        System.out.println("Polling to confirm the index has changed from the BUILDING status.");
        waitForIndex(collection, INDEX_NAME);
    }

    public static <T> boolean waitForIndex(final MongoCollection<T> collection, final String indexName) {
        long startTime = System.nanoTime();
        long timeoutNanos = TimeUnit.SECONDS.toNanos(60);
        while (System.nanoTime() - startTime < timeoutNanos) {
            Document indexRecord = StreamSupport.stream(collection.listSearchIndexes().spliterator(), false)
                    .filter(index -> indexName.equals(index.getString("name")))
                    .findAny().orElse(null);
            if (indexRecord != null) {
                if ("FAILED".equals(indexRecord.getString("status"))) {
                    throw new RuntimeException("Search index has FAILED status.");
                }
                if (indexRecord.getBoolean("queryable")) {
                    System.out.println(indexName + " index is ready to query");
                    return true;
                }
            }
            try {
                Thread.sleep(100); // busy-wait, avoid in production
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                throw new RuntimeException(e);
            }
        }
        return false;
    }
}
